source/FormationPattern.cpp

```diff
@@ -15,44 +15,124 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 
 #include "FormationPattern.h"
 
+#include "Angle.h"
 #include "DataNode.h"
 
 #include <cmath>
 
 using namespace std;
 
-namespace {
-	const Point defaultFormationPoint = Point();
-}
-
-
 
-FormationPattern::PositionIterator::PositionIterator(const FormationPattern &pattern)
-	: pattern(pattern)
+FormationPattern::PositionIterator::PositionIterator(const FormationPattern &pattern,
+		double centerBodyRadius)
+	: pattern(pattern), centerBodyRadius(centerBodyRadius)
 {
-	positionIt = pattern.positions.begin();
+	MoveToValidPositionOutsideCenterBody();
 }
 
 
 
 const Point &FormationPattern::PositionIterator::operator*()
 {
-	if(positionIt == pattern.positions.end())
-		return defaultFormationPoint;
-	return *positionIt;
+	return currentPoint;
 }
 
 
 
 FormationPattern::PositionIterator &FormationPattern::PositionIterator::operator++()
 {
-	if(positionIt != pattern.positions.end())
-		++positionIt;
+	if(!atEnd)
+	{
+		position++;
+		MoveToValidPositionOutsideCenterBody();
+	}
+
 	return *this;
 }
 
 
 
+void FormationPattern::PositionIterator::MoveToValidPositionOutsideCenterBody()
+{
+	MoveToValidPosition();
+	unsigned int maxTries = 50;
+	// Skip positions too close to the center body
+	while(!atEnd && currentPoint.Length() <= centerBodyRadius && maxTries > 0)
+	{
+		position++;
+		MoveToValidPosition();
+		maxTries--;
+	}
+	if(maxTries == 0)
+		atEnd = true;
+}
+
+
+
+void FormationPattern::PositionIterator::MoveToValidPosition()
+{
+	unsigned int lines = pattern.Lines();
+
+	// If we cannot calculate any new positions, then just return center point.
+	if(atEnd || lines < 1)
+	{
+		atEnd = true;
+		currentPoint = Point();
+		return;
+	}
+
+	unsigned int ringsScanned = 0;
+	unsigned int startingRing = ring;
+	unsigned int lineRepeatPositions = pattern.Positions(ring, line, repeat);
+
+	while(position >= lineRepeatPositions && !atEnd)
+	{
+		unsigned int patternRepeats = pattern.Repeats(line);
+		// LinePosition number is beyond the amount of positions available on the line/arc.
+		// Need to move a ring, a line/arc or a repeat-section forward.
+		if(ring > 0 && line < lines && patternRepeats > 0 && repeat < patternRepeats - 1)
+		{
+			// First check if we are on a valid line and have another repeat section.
+			++repeat;
+			position = 0;
+			lineRepeatPositions = pattern.Positions(ring, line, repeat);
+		}
+		else if(line < lines - 1)
+		{
+			// If we don't have another repeat section, then check for a next line.
+			++line;
+			repeat = 0;
+			position = 0;
+			lineRepeatPositions = pattern.Positions(ring, line, repeat);
+		}
+		else
+		{
+			// If we checked all lines and repeat sections, then go for the next ring.
+			++ring;
```

Comment:
I don' see any example formation using rings. Are they fully supported?

Reply:
Rings are used for repeat-sections, if a pattern is growing, then we need to know which iteration of the repeat section(s) we are processing. This was called "the ring" in older versions of the formations-PR, and it is thus still in use to track the iteration of the pattern.
Rings are no longer visible outside the code. The keywords to set a starting-ring manually have been removed (as part of simplification of the syntax).